### PR TITLE
Set FITS ASCII table int dtype as required by column width given by TFORMn

### LINKED ISSUE #9899: Undeclared (?) limitation of FITS ASCII table integer field width
### Description
According to the FITS standard (section 7.2.5):

"This Standard does not impose an upper limit on the number of digits of precision, nor any limit
on the range of numeric values. Software packages that read or write data according to this Standard could be limited, however, in the range of values and exponents that are supported (e.g.,
to the range that can be represented by 32-bit or 64-bit binary numbers)."

My expectation was that AstroPy would choose an underlying data-type suitable for the range of values being read, but that appears not to be the case. This file:

[i10_test.zip](https://github.com/astropy/astropy/files/4122596/i10_test.zip)

for example reads fine with the FITS reference software (fv, fitsverify, fdump etc.) but in AstroPy the values of 4294967295 are read incorrectly as -1. It looks like integer values in ASCII tables are constrained to be 32-bit signed integers - is this correct? I understand I can use offsets etc. but I couldn't see this limitation written down, so wanted to confirm.

### Expected behavior
Valid FITS files to be read as expected.

### Actual behavior
Values outside of the range of a 32-bit signed integer are incorrectly read.

### System Details
Linux-4.18.0-18-lowlatency-x86_64-with-debian-buster-sid
Python 3.6.7 | packaged by conda-forge | (default, Nov  6 2019, 16:19:42) 
[GCC 7.3.0]
Numpy 1.17.5
astropy 4.0
Scipy 1.4.1

### PR DESCRIPTION
**Pull request by dhomeier**
### Description
This pull request adds support for integer columns exceeding 32 bit precision to `fits.TableHDU`.
Previously ASCII table columns with format `I10` or wider, which might contain values `>2**31` would have been cast to `np.int32`; with this fix they are correctly read in to `np.int64`.
Fixes #9899

### CODE REVIEW

**Comment by pllim** [comment]
Please add a change log under the assigned milestone as a bug fix. Thanks!

**Comment by dhomeier** [comment]
Added some tests for the formats of newly initialised `fits.Column('TEST', 'I[n]', ascii=True)` columns and found that explicit initialisation of `recformat` was required to reproduce the previous results.
Otherwise the non-reciprocity of binary and ascii formats would create a kind of race condition, e.g. with the default `dtype` for `I` without width specification being `i4`, the default width is set to 10, but that requires `i8` to be fully represented. However the actual required width for signed `i4` is already 11 digits! Now enforcing the defaults for both, but I wonder if it would not make more sense to set the defaults to `recformat='i4'` and `width=11` right away.

**Comment by saimn** [comment]
Thanks for the PR Derek :+1: 
I'm leaving for a few days so I will review more carefully next week.

**Comment by bsipocz** [comment]
(CI failures are unrelated, please rebase after the review is done, as the travis one has been fixed already)

File: astropy/io/fits/column.py [hunk 1]
```diff
@@ -320,6 +320,10 @@ def __new__(cls, format, strict=False):
         self.format, self.width, self.precision = \
             _parse_ascii_tformat(format, strict)
 
+        # If no width has been specified, set the dtype here to default as well
+        if format == self.format:
+            self.recformat = ASCII2NUMPY[format]
```
Comment (saimn) [inline astropy/io/fits/column.py]: Why this change ? A default width should already been set by `_parse_ascii_tformat` if needed, and the `recformat` property will use this as well.

File: astropy/io/fits/fitsrec.py [hunk 2]
```diff
@@ -823,7 +823,7 @@ def _convert_ascii(self, column, field):
         """
 
         format = column.format
-        recformat = ASCII2NUMPY[format[0]]
+        recformat = getattr(column, 'dtype', ASCII2NUMPY[format[0]])
```
Comment (saimn) [inline astropy/io/fits/fitsrec.py]: Not sure about this, I think I see how the old code could be an issue, but I wonder why this was not using `column.recformat` directly ?

File: astropy/io/fits/column.py [hunk 3]
```diff
@@ -2447,9 +2451,14 @@ def _convert_ascii_format(format, reverse=False):
 
         # The following logic is taken from CFITSIO:
         # For integers, if the width <= 4 we can safely use 16-bit ints for all
-        # values [for the non-standard J format code just always force 64-bit]
-        if format == 'I' and width <= 4:
-            recformat = 'i2'
+        # values, if width >= 10 we may need to accomodate 64-bit ints.
+        # [for the non-standard J format code just always force 64-bit
+        #  - why, shouldn't it be 32-bit?]
```
Comment (saimn) [inline astropy/io/fits/column.py]: Comments are not the best place to ask questions ;). Let's leave J what it is, I think we will never know why this is supported anyway.

File: astropy/io/fits/column.py (hunk 1 above)
Comment (dhomeier) [reply astropy/io/fits/column.py]: See my comment about the [race condition](https://github.com/astropy/astropy/pull/9901#issuecomment-580208749) -`_parse_ascii_tformat` will first set the default `width` to `10`, but then infer a required `recformat` of `i8` for that. But `I10` is not even sufficient to fully represent `i4`, which seems problematic when creating a new column to be written as ascii table. `test_column_format_interpretation` created the default `fits.Column('TEST', 'I', ascii=True)` with `i4`, although that was not actually tested. But I decided to leave it at that for now.
OTOH when creating a new column to be written to a file, we might not care about the exact `recformat`, since it comes out `ASCII`-formatted in the end anyway. From that point it might be preferable to set `i8` and `I11` as the defaults, which would support the full range of signed `int32` for writing and provide extra internal precision for any operations one might perform on the column data before writing. But writing as `I11` per default would of course mean changing output.

File: astropy/io/fits/fitsrec.py (hunk 2 above)
Comment (dhomeier) [reply astropy/io/fits/fitsrec.py]: Should better document my changes to myself, it's already too long ago...
The core of the issue was, I think, that `_convert_ascii` might be called on a column that does not yet have a `recformat` set (I would have thought that `Table.read()` would already read in the table with `data` as `np.array` of the appropriate `dtype`, but it turned out it is initially stored as `str` columns, which are only converted with `_convert_ascii` “on demand”).
In any case `getattr(column, 'recformat')` still reads the example file from #9899 incorrectly as `int32`, although it passes all the tests, so I guess there is another test to be added for that...

Comment (saimn) [reply astropy/io/fits/fitsrec.py]: Hum, this is why io.fits code is hard to change :D 
I also forgot the details but in the past I tried to replace this line with `format.recformat` (https://github.com/astropy/astropy/pull/5362/commits/0fc487e8a8b43ed1cfb0f2e4a53566021042171d) and reverted this in the commit after because we chose to read all float values as double (#5362, #5353).
Maybe we should just do the same for integers and always read as int64 ? (cc @MSeifert04 if you have a better memory than me ;)).

File: astropy/io/fits/column.py (hunk 3 above)
Comment (dhomeier) [reply astropy/io/fits/column.py]: It was more of a rhetorical question, maybe hoping someone would someday remember what extension this was based on. The `J` format definition first entered as part of a fairly [massive merge](https://github.com/astropy/astropy/commit/346517f) by @embray, does not  look like it's possible to trace it back to the actual origin.

File: astropy/io/fits/fitsrec.py (hunk 2 above)
Comment (dhomeier) [reply astropy/io/fits/fitsrec.py]: Ah, `column.recformat` did not work, because the `_AsciiColumnFormat` method is just a wrapper around `_convert_ascii_format` and will thus determine `recformat` anew from `format` (maybe it should first check if `self.recformat` or `self.format.recformat` already exist?).

File: astropy/io/fits/column.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/column.py]: I spend some time to remember myself this code, and sorry I don't understand what you mean here.
I don't see a case where your change would fix something that is done by `_parse_ascii_tformat` ?

Comment (saimn) [reply astropy/io/fits/column.py]: That said I think most issues are fixed by changing the default `recformat` to `i8` in `ASCII2NUMPY`. And I think we should also change the default to `I11` as well.

File: astropy/io/fits/fitsrec.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/fitsrec.py]: yes sorry that's `format.recformat`, but if we change the default value that does not matter anymore.

> So I think changing results by making int64 always the default should be deferred to a PR for a later milestone. In terms of memory efficiency it would probably be preferable to stick with the current tailored formats (although you could say if memory usage becomes an issue, ascii formats are hardly a sensible choice anyway).

Memory usage should not be an issue, in particular for ascii files. I prefer that we fix the bug and produce valid results, so I think it is fine with the next bugfix release.

File: astropy/io/fits/column.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/column.py]: Sometimes it is possible to get more information in pyfits' repository (https://github.com/spacetelescope/PyFITS/commit/9096a66efa99b5111aeaf7b3f9e2bb47a1c467f8), but here, unless Erik remembers...

Comment (pllim) [reply astropy/io/fits/column.py]: If you get desperate, I think the SVN repo is archived somewhere... but I'd rather not waste time digging if this is purely rhetorical.

**Comment by bsipocz** [comment]
Please rebase to have another run for CI and conflict resolution.

**Comment by saimn** [comment]
So, what do we choose here?
I still think that it is better change the default values in `ASCII2NUMPY` and `ASCII_DEFAULT_WIDTHS` now, to use int64 by default, as we did for float/double (see discussion in #5353), instead of adding more hacks to something that is already quite complicated. 
It is also the safest option, to limit the risk of overflow. If someone needs a more memory efficient then it is easy to create an int32 column, but by default we should use the maximum size to avoid surprises (and bug reports ;)).
@MSeifert04, thoughts ?

**Comment by saimn** [comment]
@dhomeier - Could you rebase this branch and move the changelog entry to 4.0.2 ?

File: astropy/io/fits/fitsrec.py (hunk 2 above)
Comment (embray) [reply astropy/io/fits/fitsrec.py]: Keeping with int types tailored to the column makes sense.  For example, sometimes there are "boolean" columns that only contain 0 or 1--it makes to sense to waste 8 bytes per row on such columns, when an 8-bit int can be used instead.

File: astropy/io/fits/column.py (hunk 3 above)
Comment (embray) [reply astropy/io/fits/column.py]: I'm trying to recall the history of the non-standard "J" format for ASCII columns.  It's been so long I can't really remember, but I believe there were some files in the wild that actually used this.  I don't remember where they came from.

Comment (embray) [reply astropy/io/fits/column.py]: Looking at the change @pllim pointed out, its existence in the code definitely predated me working on it.  It's possible I just kept it for backwards-compatibility.  I don't recall anymore whether it's actually been used.  I wonder if anyone else knows...

Comment (embray) [reply astropy/io/fits/column.py]: Hmm, maybe I did add it.  It first appears in the git repo at https://github.com/spacetelescope/PyFITS/commit/0e5a8a213e1f3ab919e2e577176719df3dc622cb, which according to the commit message was somehow carried over from SVN, but the commit message in git doesn't clarify further.  I might have a copy of the svn repo somewhere, let me check...

File: astropy/io/fits/column.py [hunk 4]
```diff
@@ -2470,9 +2474,13 @@ def _convert_ascii_format(format, reverse=False):
...
         # For integers, if the width <= 4 we can safely use 16-bit ints for all
+        # values, if width >= 10 we may need to accomodate 64-bit ints.
         # values [for the non-standard J format code just always force 64-bit]
-        if format == 'I' and width <= 4:
-            recformat = 'i2'
+        if format == 'I':
+            if width <= 4:
+                recformat = 'i2'
+            elif width > 9:
+                recformat = 'i8'
```
Comment (embray) [inline astropy/io/fits/column.py:2483]: I wonder if there should be a warning in the case of widths beyond that supported by a 64-bit integer.  Of course the problem with this format is you can have a column with a format like `I50` but that contains mostly whitespace in each entry, while the actual values are much smaller.

**Comment by embray** [comment]
A somewhat related issue is #3430.  The idea there would be to use the `io.ascii` machinery to parse FITS ASCII tables, which would just become another ASCII table format. IIUC `io.ascii` just makes all int columns int64 by default, though for FITS tables I might still include some logic to use the minimum needed int type.

**Review by embray — APPROVED** [review]
I see no reason to further delay this.  It clearly resolves this issue without affecting backwards compatibility.  The fix makes sense and is well tested.
